File: Scala/graph_algorithms/src/main/scala/decompositions/ConnectedComponents.scala
package decompositions

object ConnectedComponents {
  import scala.annotation.tailrec

  type Node = Int
  type Component = List[Node]

  final case class Edge(from: Node, to: Node)

  class Graph(edgeList: List[Edge], val nrNodes: Int) {
    val adjacencyList: Map[Node, List[Node]] = Graph.createAdjacencyList(edgeList)

    lazy val connectedComponents: List[Component] = calcConnectedComponents()

    private def calcConnectedComponents(): List[Component] = {
      val visitStarted: Array[Int] = Array.fill(nrNodes)(0)
      val visitEnded: Array[Int] = Array.fill(nrNodes)(0)
      val previsitId: Iterator[Int] = Iterator.from(1)
      val postvisitId: Iterator[Int] = Iterator.from(1)

      def isNodeVisited(node: Node): Boolean = visitStarted(node - 1) > 0

      def findUnvisitedNeighbour(node: Node): Option[Node] =
        adjacencyList.getOrElse(node, Nil).find(!isNodeVisited(_))

      @tailrec
      def findComponents(nodes: List[Node], components: List[Component]): List[Component] = nodes match {
        case Nil => components
        case node :: remainingNodes =>
          if (isNodeVisited(node)) findComponents(remainingNodes, components)
          else {
            val currentComponent: Component = explore(node)
            findComponents(remainingNodes, currentComponent :: components)
          }
      }

      def explore(startingNode: Node): Component = {
        @tailrec
        def traverseComponent(previsitStack: List[Node], component: Component): Component = previsitStack match {
          case Nil => component
          case node :: restOfStack =>
            findUnvisitedNeighbour(node) match {
              case Some(neighbour) =>
                visitStarted(neighbour - 1) = previsitId.next()
                traverseComponent(neighbour :: previsitStack, neighbour :: component)
              case None =>
                visitEnded(node - 1) = postvisitId.next()
                traverseComponent(restOfStack, component)
            }
          }

        visitStarted(startingNode - 1) = previsitId.next()
        traverseComponent(List(startingNode), List(startingNode))
      }

      findComponents((1 to nrNodes).toList, Nil)
    }
  }

  object Graph {
    private def createAdjacencyList(edgeList: List[Edge]): Map[Node, List[Node]] = {
      val directedEdges: List[Edge] = edgeList.flatMap{ case e @ Edge(a, b) => List(e, Edge(b, a)) }
      directedEdges.groupBy(_.from).mapValues(_.map(_.to))
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(nrNodes, nrEdges): List[Int] = convertToIntList(reader.next())
    val edgeList: List[Edge] = (for { _ <- 0 until nrEdges } yield {
      val List(a, b): List[Node] = convertToIntList(reader.next())
      Edge(a, b)
    }).toList
    val graph: Graph = new Graph(edgeList, nrNodes)
    val components: List[Component] = graph.connectedComponents
    println(components.length)
  }
}

File: OCaml/graph_algorithms/decompositions/connected_components.ml
type edge = { left_end : int; right_end : int }

module IntDict = Hashtbl.Make(
    struct
        type t = int
        let equal = Pervasives.(=)
        let hash = Hashtbl.hash
    end)

module type GraphSig =
    sig
        type t
        type component = int list
        val create_graph : edge list -> int -> t
        val get_neighbours : t -> int -> int list
        val connected_components : t -> component list
    end

module Graph : GraphSig =
    struct
        type t = { adjacency_list : int list IntDict.t; nr_nodes : int }
        type component = int list

        let reverse { left_end = a; right_end = b } = { left_end = b; right_end = a }

        let range a b =
            let rec loop acc left =
                if left >= b then List.rev acc
                else loop (left :: acc) (left + 1)
            in loop [] a

        let get_neighbours { adjacency_list } node =
            match IntDict.find_opt adjacency_list node with
                | None -> []
                | Some neighbours -> neighbours

        let create_graph edges nr_nodes =
            let adjacency_list = IntDict.create 100 in
            let add_edge { left_end; right_end } =
                let neighbours = get_neighbours { adjacency_list; nr_nodes } left_end in
                IntDict.replace adjacency_list left_end (right_end :: neighbours) in
            List.iter add_edge edges;
            List.iter (function e -> add_edge @@ reverse e) edges;
            { adjacency_list; nr_nodes }

        let connected_components ({ nr_nodes } as graph) =
            let visit_started = Array.make nr_nodes 0 in
            let visit_ended = Array.make nr_nodes 0 in
            let previsit_id = ref 1 in
            let postvisit_id = ref 1 in
            let is_node_visited node = visit_started.(node - 1) > 0 in
            let find_unvisited_neighbour node =
                List.find_opt (function elem -> not (is_node_visited elem)) (get_neighbours graph node) in
            let explore starting_node =
                let rec traverse_component previsit_stack component = match previsit_stack with
                    | [] -> component
                    | node :: rest_of_stack -> match find_unvisited_neighbour node with
                        | Some neighbour ->
                            visit_started.(neighbour - 1) <- !previsit_id;
                            previsit_id := !previsit_id + 1;
                            traverse_component (neighbour :: previsit_stack) (neighbour :: component)
                        | None ->
                            visit_ended.(node - 1) <- !postvisit_id;
                            postvisit_id := !postvisit_id + 1;
                            traverse_component rest_of_stack component in
                visit_started.(starting_node - 1) <- !previsit_id;
                previsit_id := !previsit_id + 1;
                traverse_component [starting_node] [starting_node] in
            let rec find_components nodes components = match nodes with
                | [] -> components
                | node :: remaining_nodes ->
                    if is_node_visited node then find_components remaining_nodes components
                    else
                        let current_component = explore node in
                        find_components remaining_nodes (current_component :: components)
            in find_components (range 1 (nr_nodes + 1)) []
    end


let read_two_ints() = Scanf.sscanf (read_line ()) "%d %d" (fun x y -> x, y)


let read_edges nr_edges =
    let rec loop edges k =
        if k = nr_edges then edges
        else
            let left_end, right_end = read_two_ints() in
            loop ({ left_end; right_end } :: edges) (k + 1)
    in loop [] 0


let () =
    let nr_nodes, nr_edges = read_two_ints() in
    let edges = read_edges nr_edges in
    let graph = Graph.create_graph edges nr_nodes in
    let components = Graph.connected_components graph in
    print_int (List.length components);
    print_newline()
